### PR TITLE
feat: add Twist group recipients support

### LINKED ISSUE #176: Add tool to fetch group id
I tried creating a Twist automation and wanted it to only notify a specific group, but it couldn't get the group id automatically.

`"26Q2|Prod: Automation" is likely a group, not a user — but the get-users tool only searches users. Unfortunately there's no get-groups tool available in the Twist connector, so I can't look up the group ID automatically.
`

### PR DESCRIPTION
**Pull request by hfauq**
## Summary

Agents can now resolve Twist group IDs and notify those groups from the thread-writing tools. This adds a read-only `get-groups` tool, then extends `create-thread` and thread replies to accept `groups` alongside user `recipients`.

Thread replies also now default to notifying everyone who has interacted with the thread instead of relying on Twist's broad channel default. The mutation tools echo requested `recipients` and `groups`, plus the thread-reply default recipient mode, in structured output so automations can confirm notification behavior.

## Details

- Adds `get-groups` with optional `groupIds` and `searchText` filters.
- Registers and exports the new tool through the MCP server, package index, direct runner, and tool annotation coverage.
- Passes `groups` through `create-thread`.
- Passes `groups` through `reply` for `targetType: "thread"` while leaving conversation replies unchanged.
- Sends `EVERYONE_IN_THREAD` for thread replies when no explicit users or groups are provided.
- Updates MCP instructions and output schemas so agents know to resolve group names with `get-groups` before writing.

Closes Doist/twist-ai#176.

## Test Plan

- `npm test`
- `npm run type-check`
- `npm run format:check`
- `npm run build`

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)
GPT-5

### CODE REVIEW

File: src/tools/reply.ts [hunk 1]
```diff
@@ -18,6 +18,12 @@ const ArgsSchema = {
         .describe(
             'Optional array of user IDs to notify (only for thread replies). If omitted, Twist defaults to notifying all current members of the channel. Add specific user IDs to limit or expand notifications beyond current channel members.',
         ),
+    groups: z
```
Comment (doistbot) [inline src/tools/reply.ts:29]: [P2] `groups` is now part of the public `reply` input for every target type, but conversation replies still ignore it and return `success: true`. That means a caller can request group notifications on a conversation message and get a successful response even though no group was notified. Please reject `groups` when `targetType === "conversation"` (or otherwise make the schema/runtime validation enforce the thread-only constraint) instead of silently dropping it.

File: src/tools/get-groups.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    id: number
+    name: string
+    description?: string
+    workspaceId: number
+    userIds: number[]
+    memberCount: number
+    version: number
+}
+
+type GetGroupsStructured = Record<string, unknown> & {
```
Comment (doistbot) [inline src/tools/get-groups.ts]: [P2] This file defines `GroupData`/`GetGroupsStructured` even though the same output contract is already captured by `GetGroupsOutputSchema` and exported as `GetGroupsOutput` in `output-schemas.ts`. Reusing the schema-derived type here would remove a second copy of the shape and prevent future drift when group output fields change.

File: src/tools/reply.ts [hunk 3]
```diff
@@ -29,6 +35,8 @@ type ReplyStructured = {
     content: string
     created: string
     replyUrl: string
+    recipients?: number[]
```
Comment (doistbot) [inline src/tools/reply.ts]: [P2] Adding these fields extends a local `ReplyStructured` type that duplicates `ReplyOutputSchema`. `output-schemas.ts` already exports `ReplyOutput`, and `create-thread` uses that pattern nearby. Reusing the schema-derived type here would avoid having to update two separate reply output definitions every time this tool's structured payload changes.

**Review by doistbot — COMMENTED** [review]
Thank you for adding support for Twist group recipients by introducing the `get-groups` tool and extending the thread tools to handle group notifications. These updates will significantly improve the agents' ability to effectively notify the right audience and verify their intended targets. To polish the implementation, there are a few adjustments needed around reusing existing schema-derived types, improving type safety at the SDK boundary, and ensuring that group targets are explicitly rejected for conversation replies. Additionally, optimizing the `get-groups` payload to omit user IDs and user-generated descriptions by default, alongside expanding test coverage for empty array edge cases, will help ensure a robust and secure integration.

<sub>[Share Feedback](https://github.com/Doist/doistbot-app/issues/new?template=review-feedback.yml&title=Review+feedback+for+Doist%2Ftwist-ai&pr_url=https%3A%2F%2Fgithub.com%2FDoist%2Ftwist-ai%2Fpull%2F188&datadog_url=https%3A%2F%2Fapp.datadoghq.com%2Flogs%3Fquery%3Dservice%253A%2528doistbot-app-sandbox%2520OR%2520doistbot-app%2529%2520%2540meta.deliveryId%253A520d7ef0-4891-11f1-881e-9fdf80dda107%26agg_m%3Dcount%26agg_m_source%3Dbase%26agg_t%3Dcount%26clustering_pattern_field_path%3Dmessage%26cols%3Dhost%252Cservice%26fromUser%3Dtrue%26messageDisplay%3Dinline%26refresh_mode%3Dsliding%26storage%3Dhot%26stream_sort%3Ddesc%26viz%3Dstream%26live%3Dtrue) • [Review Logs](https://app.datadoghq.com/logs?query=service%3A%28doistbot-app-sandbox%20OR%20doistbot-app%29%20%40meta.deliveryId%3A520d7ef0-4891-11f1-881e-9fdf80dda107&agg_m=count&agg_m_source=base&agg_t=count&clustering_pattern_field_path=message&cols=host%2Cservice&fromUser=true&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&live=true)</sub>

File: src/tools/reply.ts [hunk 4]
```diff
@@ -39,18 +47,20 @@ const reply = {
...
 
         if (targetType === 'thread') {
-            const comment = await client.comments.createComment({
+            const commentArgs = {
                 threadId: targetId,
                 content,
                 recipients,
-            })
+                groups,
+            } as Parameters<typeof client.comments.createComment>[0] & { groups?: number[] }
```
Comment (doistbot) [inline src/tools/reply.ts]: [P2] The cast here sidesteps the SDK type for `comments.createComment` to force in `groups`. That removes compiler protection exactly at the API boundary this feature depends on, so future SDK/request-shape drift will not be caught. Prefer updating/augmenting the Twist SDK request type or introducing a typed local wrapper instead of asserting an extra field onto the payload.

File: src/tools/get-groups.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+        const structuredContent: GetGroupsStructured = {
+            type: 'get_groups',
+            workspaceId,
+            groups: filteredGroups.map((group) => ({
+                id: group.id,
+                name: group.name,
+                ...(group.description && { description: group.description }),
+                workspaceId: group.workspaceId,
+                userIds: group.userIds,
```
Comment (doistbot) [inline src/tools/get-groups.ts]: [P2] Including every group's full `userIds` array here makes `get-groups` responses scale with total group membership, not just group count. Because `getToolOutput()` also mirrors structured content into JSON for legacy clients, this payload gets duplicated on the wire. For the common "resolve group IDs before notifying" path, prefer omitting member IDs by default or making membership expansion opt-in.

File: src/tools/__tests__/create-thread.test.ts [hunk 6]
```diff
@@ -80,9 +86,75 @@ describe(`${CREATE_THREAD} tool`, () => {
...
             })
 
             expect(extractTextContent(result)).toMatchSnapshot()
+
+            const structuredContent = extractStructuredContent(result)
+            expect(structuredContent.recipients).toEqual([TEST_IDS.USER_1, TEST_IDS.USER_2])
+            expect(structuredContent).not.toHaveProperty('groups')
+        })
+
+        it('should create a thread with groups', async () => {
```
Comment (doistbot) [inline src/tools/__tests__/create-thread.test.ts:99]: [P2] These new cases only cover populated `groups`, so the behavior for `groups: []` is still unpinned. For a newly added optional notification field, an explicit empty-array test is high value because it locks down whether the tool preserves an intentional "no group notifications" request in both the API call and structured output instead of silently treating it like omission.

File: src/tools/__tests__/reply.test.ts [hunk 7]
```diff
@@ -82,9 +84,67 @@ describe(`${REPLY} tool`, () => {
...
             })
 
             expect(extractTextContent(result)).toMatchSnapshot()
+
+            const structuredContent = extractStructuredContent(result)
+            expect(structuredContent.recipients).toEqual([TEST_IDS.USER_1, TEST_IDS.USER_2])
+            expect(structuredContent).not.toHaveProperty('groups')
+        })
+
+        it('should post a comment with groups', async () => {
```
Comment (doistbot) [inline src/tools/__tests__/reply.test.ts:99]: [P2] The thread-reply coverage for the new `groups` field is missing the empty-array edge case. Please add a `groups: []` test here as well, so the suite documents whether a caller can explicitly send an empty group target list and ensures that behavior doesn't regress in the forwarded `createComment` args or returned structured output.

File: src/tools/get-groups.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        lines.push('')
+
+        if (filteredGroups.length === 0) {
+            lines.push('No groups found.')
+        } else {
+            for (const group of filteredGroups) {
+                lines.push(`## ${group.name}`)
+                lines.push(`**ID:** ${group.id}`)
+                lines.push(`**Members:** ${group.userIds.length}`)
+                if (group.description) {
```
Comment (doistbot) [inline src/tools/get-groups.ts]: [P2] This tool is described as an ID-discovery helper, but including `group.description` here (and again in structured output below) forwards arbitrary user-generated Twist text into the LLM context even though it is not needed to resolve group IDs. The internal AI tools standard requires minimizing data sent to AI systems and avoiding user-generated text unless necessary. Please drop descriptions from the output, or make them an explicit opt-in field. https://handbook.doist.com/doc/standard-internal-ai-tools-zU0fqnhpmC

File: src/tools/reply.ts (hunk 3 above)
Comment (hfauq) [reply src/tools/reply.ts]: Addressed in 0454817: `ReplyStructured` now reuses the exported schema-derived `ReplyOutput` type instead of duplicating the structured output contract locally.

File: src/tools/create-thread.ts [hunk 9]
```diff
@@ -76,6 +83,8 @@ const createThread = {
             creator: thread.creator,
             created: created.toISOString(),
             threadUrl,
+            ...(recipients && { recipients }),
```
Comment (scottlovegrove) [inline src/tools/create-thread.ts]:
```suggestion
            ...(recipients ? { recipients } : {}),
```

File: src/tools/create-thread.ts [hunk 10]
```diff
@@ -76,6 +83,8 @@ const createThread = {
...
+            ...(groups && { groups }),
```
Comment (scottlovegrove) [inline src/tools/create-thread.ts]:
```suggestion
            ...(groups ? { groups } : {}),
```

File: src/tools/get-groups.ts [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
+    async execute(args, client) {
+        const { workspaceId, groupIds, searchText } = args
+
+        const workspaceGroups = await client.groups.getGroups(workspaceId)
+        const requestedGroupIds = groupIds && groupIds.length > 0 ? new Set(groupIds) : undefined
+
+        const groups = requestedGroupIds
+            ? workspaceGroups.filter((group) => requestedGroupIds.has(group.id))
+            : workspaceGroups
```
Comment (scottlovegrove) [inline src/tools/get-groups.ts]: If we have a list of requested group IDs, we can just batch the calls and fetch all requested at once, rather than fetch _all_ workspace groups and compare. 

If no requested groups array is provided, _then_ we can just make the call to get all groups.

File: src/tools/reply.ts [hunk 12]
```diff
@@ -16,40 +23,63 @@ const ArgsSchema = {
...
+type CommentsClientWithGroupRecipients = {
+    createComment(args: CreateThreadReplyCommentArgs): Promise<Comment>
+}
+
+async function createThreadReplyComment(
+    client: TwistApi,
+    args: CreateThreadReplyCommentArgs,
+): Promise<Comment> {
+    // The Twist API accepts comment groups, but twist-sdk has not typed that field yet.
+    const comments = client.comments as unknown as CommentsClientWithGroupRecipients
```
Comment (scottlovegrove) [inline src/tools/reply.ts]: @hfauq Can you please open an issue at https://github.com/Doist/twist-sdk-typescript/issues to get this corrected in the SDK and a follow up PR can correct this erroneous typing.

File: src/tools/reply.ts [hunk 13]
```diff
@@ -1,11 +1,18 @@
...
 import { z } from 'zod'
 import { getToolOutput } from '../mcp-helpers.js'
 import type { TwistTool } from '../twist-tool.js'
-import { ReplyOutputSchema } from '../utils/output-schemas.js'
-import { type ReplyTargetType, ReplyTargetTypeSchema } from '../utils/target-types.js'
+import { type ReplyOutput, ReplyOutputSchema } from '../utils/output-schemas.js'
+import { ReplyTargetTypeSchema } from '../utils/target-types.js'
 import { ToolNames } from '../utils/tool-names.js'
 
+const DEFAULT_THREAD_REPLY_RECIPIENTS = 'EVERYONE_IN_THREAD' as const
```
Comment (scottlovegrove) [inline src/tools/reply.ts:15]: Hmm, this should probably live in the Twist SDK too. Again, can you create an issue to port this to the SDK please.

File: src/tools/reply.ts [hunk 14]
```diff
@@ -108,6 +138,11 @@ const reply = {
             content,
             created: created.toISOString(),
             replyUrl,
+            ...(targetType === 'thread' && recipients && { recipients }),
+            ...(targetType === 'thread' &&
+                !recipients &&
+                !groupsToNotify && { recipientMode: DEFAULT_THREAD_REPLY_RECIPIENTS }),
+            ...(targetType === 'thread' && groupsToNotify && { groups: groupsToNotify }),
```
Comment (scottlovegrove) [inline src/tools/reply.ts]: This should be terniary checks.

File: src/utils/output-schemas.ts [hunk 15]
```diff
@@ -243,6 +243,25 @@ export const GetUsersOutputSchema = z.object({
...
+export const GetGroupsOutputSchema = z.object({
+    type: z.literal('get_groups'),
+    workspaceId: z.number(),
+    groups: z.array(
+        z.object({
+            id: z.number(),
+            name: z.string(),
+            workspaceId: z.number(),
+            memberCount: z.number(),
+            version: z.number(),
```
Comment (scottlovegrove) [inline src/utils/output-schemas.ts]: Not sure we need the version here 🤔

File: src/tools/create-thread.ts (hunk 9 above)
Comment (hfauq) [reply src/tools/create-thread.ts]: Applied in ff139ff: the `recipients` structured-output spread now uses the explicit ternary form.

File: src/tools/get-groups.ts (hunk 11 above)
Comment (hfauq) [reply src/tools/get-groups.ts]: Fixed in ff139ff: when `groupIds` are provided, `get-groups` now dedupes them and batch-fetches those IDs via `getGroup(..., { batch: true })`. It only calls `getGroups(workspaceId)` when no non-empty `groupIds` array is provided, and the tests assert that path.

File: src/tools/reply.ts (hunk 12 above)
Comment (hfauq) [reply src/tools/reply.ts]: Opened the SDK follow-up as https://github.com/Doist/twist-sdk-typescript/issues/124 and linked it from the local wrapper in ff139ff.

**Comment by hfauq** [comment]
@scottlovegrove Thanks for the snappy review. Can you have another look?

Next steps:
- I've also opened two issues in the SDK repo as you suggested – will you take care of them? Are they dependencies to then merge this PR here? 
- We need to handle those changes on the Automations' Twist connector, will you be doing this?

**Comment by scottlovegrove** [comment]
> * I've also opened two issues in the SDK repo as you suggested – will you take care of them?

Yeah, I'll take care of those. 

> * Are they dependencies to then merge this PR here?

Nah

> * We need to handle those changes on the Automations' Twist connector, will you be doing this?

Once this is merged and deployed, there's nothing to do with Automations, it will just pick up the changes automatically.

**Review by scottlovegrove — APPROVED** [review]

**Comment by hfauq** [comment]
Awesome, merging.

> Once this is merged and deployed, there's nothing to do with Automations, it will just pick up the changes automatically.

🔥 Where can I follow the update? I'd like to test the automation with groups to validate!

**Comment by scottlovegrove** [comment]
@hfauq This has been deployed now, you should be able to try and create an automation with it.

**Comment by hfauq** [comment]
@scottlovegrove I ran into an issue https://github.com/Doist/twist-ai/issues/189
